framework/src/meshgenerators/CoarseElementIDFixup.C

```diff
@@ -0,0 +1,228 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "CoarseElementIDFixup.h"
+
+#include "libmesh/elem.h"
+#include "libmesh/mesh_serializer.h"
+
+registerMooseObject("MooseApp", CoarseElementIDFixup);
+
+InputParameters
+CoarseElementIDFixup::validParams()
+{
+  InputParameters params = MeshGenerator::validParams();
+  params.addRequiredParam<MeshGeneratorName>("input", "The mesh we want to modify");
+  params.addRequiredParam<std::string>("source_element_id", "The element ID to be fixed up");
+  params.addParam<bool>("verbose", false, "True to show what fine elements get ID reassignment");
+  params.addParam<bool>(
+      "order_ids_with_centroid",
+      false,
+      "True to reassign elements IDs based on the centroid of all elements with the same IDs");
+  params.addClassDescription("Check and change element IDs so that all fine elements with the same "
+                             "element ID are connected through sides.");
+  return params;
+}
+
+CoarseElementIDFixup::CoarseElementIDFixup(const InputParameters & params)
+  : MeshGenerator(params), _input(getMesh("input")), _verbose(getParam<bool>("verbose"))
+{
+}
+
+std::unique_ptr<MeshBase>
+CoarseElementIDFixup::generate()
+{
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+
+  auto src_id_name = getParam<std::string>("source_element_id");
+  if (!mesh->has_elem_integer(src_id_name))
+    mooseError("The source element ID does not exist on the input mesh");
+
+  // need to serialize the mesh for fixup
+  MeshSerializer tm(*mesh);
```

Comment:
this makes it loose a ton of appeal

Reply:
No easy way to get around this and fortunately this only happen during mesh generation stage in memory and the coarse mesh is typically not costly.

Reply:
this is not the coarse mesh though, it s the fine mesh

Reply:
Oh, you are right. I was thinking the other serialization. Doing the parallel fixup is extremely difficult. This is somewhat related with the `prepare_for_use` function. In case of distributed mesh, if the input mesh has not been distributed, here will be just a null operation I think. If the input mesh is a real distributed mesh generator, yes, this operation can be costly. We did not immediately call prepare_for_use at the end of this generator because I think MOOSE will call later anyway.